File: src/types/CombatTechnique_Ranged.ts
/**
 * @main RangedCombatTechnique
 */

import { TypeConfig } from "../typeConfig.js"
import { todo } from "../validation/builders/integrity.js"
import { validateEntityFileName } from "../validation/builders/naming.js"
import { createSchemaValidator } from "../validation/builders/schema.js"
import { getFIlenamePrefixAsNumericId } from "../validation/filename.js"
import { ImprovementCost } from "./_ImprovementCost.js"
import { LocaleMap } from "./_LocaleMap.js"
import { AttributeReference } from "./_SimpleReferences.js"
import { Errata } from "./source/_Erratum.js"
import { PublicationRefs } from "./source/_PublicationRef.js"

/**
 * @title Ranged Combat Technique
 */
export type RangedCombatTechnique = {
  /**
   * The ranged combat technique's identifier. An unique, increasing integer.
   * @integer
   * @minimum 1
   */
  id: number

  /**
   * Special rules for the combat technique that apply to all weapons in this
   * category.
   */
  special: RangedCombatTechniqueSpecialRules

  /**
   * The primary attribute(s).
   * @minItems 1
   * @maxItems 2
   * @uniqueItems
   */
  primary_attribute: AttributeReference[]

  /**
   * The *Breaking Point Rating* of the respective combat technique.
   * @integer
   * @minimum 1
   */
  breaking_point_rating: number

  improvement_cost: ImprovementCost

  src: PublicationRefs

  /**
   * All translations for the entry, identified by IETF language tag (BCP47).
   */
  translations: LocaleMap<RangedCombatTechniqueTranslation>
}

/**
 * Special rules for the combat technique that apply to all weapons in this
 * category.
 */
export type RangedCombatTechniqueSpecialRules = {
  has_ammunition: boolean
}

export type RangedCombatTechniqueTranslation = {
  /**
   * The name of the condition.
   * @minLength 1
   */
  name: string

  /**
   * Additional rules for the condition, if applicable.
   * @markdown
   * @minLength 1
   */
  special?: string

  errata?: Errata
}

export const config: TypeConfig<RangedCombatTechnique> = {
  name: "RangedCombatTechnique",
  id: getFIlenamePrefixAsNumericId,
  integrityValidator: todo("RangedCombatTechnique"),
  schemaValidator: createSchemaValidator(import.meta.url),
  fileNameValidator: validateEntityFileName,
}

File: Sources/OptolithDatabaseSchema/GeneratedTypes/CombatTechnique_Ranged.swift
//
//  CombatTechnique_Ranged.swift
//  OptolithDatabaseSchema
//

import DiscriminatedEnum

public struct RangedCombatTechnique: LocalizableEntity {
    /// The ranged combat technique's identifier. An unique, increasing integer.
    public let id: Int
    
    /// Special rules for the combat technique that apply to all weapons in this category.
    public let special: RangedCombatTechniqueSpecialRules
    
    /// The primary attribute(s).
    public let primaryAttribute: [AttributeReference]
    
    /// The *Breaking Point Rating* of the respective combat technique.
    public let breakingPointRating: Int
    
    public let improvementCost: ImprovementCost
    
    public let src: PublicationRefs
    
    /// All translations for the entry, identified by IETF language tag (BCP47).
    public let translations: LocaleMap<RangedCombatTechniqueTranslation>

    public init(id: Int, special: RangedCombatTechniqueSpecialRules, primaryAttribute: [AttributeReference], breakingPointRating: Int, improvementCost: ImprovementCost, src: PublicationRefs, translations: LocaleMap<RangedCombatTechniqueTranslation>) {
        self.id = id
        self.special = special
        self.primaryAttribute = primaryAttribute
        self.breakingPointRating = breakingPointRating
        self.improvementCost = improvementCost
        self.src = src
        self.translations = translations
    }    
    
    private enum CodingKeys: String, CodingKey {
        case id = "id"
        case special = "special"
        case primaryAttribute = "primary_attribute"
        case breakingPointRating = "breaking_point_rating"
        case improvementCost = "improvement_cost"
        case src = "src"
        case translations = "translations"
    }
}

/// Special rules for the combat technique that apply to all weapons in this category.
public struct RangedCombatTechniqueSpecialRules: EntitySubtype {
    public let hasAmmunition: Bool

    public init(hasAmmunition: Bool) {
        self.hasAmmunition = hasAmmunition
    }    
    
    private enum CodingKeys: String, CodingKey {
        case hasAmmunition = "has_ammunition"
    }
}

public struct RangedCombatTechniqueTranslation: EntitySubtype {
    /// The name of the condition.
    public let name: String
    
    /// Additional rules for the condition, if applicable.
    public let special: String?
    
    public let errata: Errata?

    public init(name: String, special: String? = nil, errata: Errata? = nil) {
        self.name = name
        self.special = special
        self.errata = errata
    }
}
